File: examples/server/02_credentials/02_exchanging_credentials/presentation_request/go/main.go
// Package main demonstrates credential exchange through presentation requests.
//
// This builds on the credential issuance examples to show how the created
// credentials form the foundation for credential exchange workflows.
//
// 📚 Prerequisites: Complete the basic credential issuance examples first:
// • ../basic/main.go - Foundation credential creation
// • ../multi_claim/main.go - Multiple claims in credentials
//
// This example shows:
// - How issued credentials are stored and organized for exchange
// - Conceptual workflow of credential presentation requests
// - Creating verifiable presentations from credentials
// - Foundation patterns for building exchange applications
//
// 🎯 What you'll learn:
// • How credential issuance enables exchange scenarios
// • Credential storage and organization patterns
// • Presentation creation from issued credentials
// • Foundation concepts for real exchange applications
//
// 💡 This demonstrates the BRIDGE between issuance and exchange!
// For actual networking and discovery, see advanced tutorials.
package main

import (
	"crypto/rand"
	"crypto/sha256"
	"fmt"
	"log"
	"time"

	"github.com/joinself/self-go-sdk/account"
	"github.com/joinself/self-go-sdk/credential"
	"github.com/joinself/self-go-sdk/keypair/signing"
)

// ExchangeParty represents a party that can issue and exchange credentials
type ExchangeParty struct {
	name        string
	account     *account.Account
	credentials map[string]*credential.VerifiableCredential
}

func main() {
	fmt.Println("Credential Exchange Demo")
	fmt.Println("========================")

	// Create parties for the exchange scenario
	issuer, holder := createExchangeParties()
	defer issuer.account.Close()
	defer holder.account.Close()

	// Issue credentials for exchange
	issueCredentialsForExchange(issuer, holder)

	// Demonstrate exchange workflow
	demonstrateExchangeWorkflow(holder)

	fmt.Println("✅ Exchange demo completed!")
}

// generateStorageKey creates a cryptographically secure 32-byte key
func generateStorageKey(seed string) []byte {
	key := make([]byte, 32)
	if _, err := rand.Read(key); err != nil {
		h := sha256.Sum256([]byte(fmt.Sprintf("self-sdk-%s-%d", seed, time.Now().UnixNano())))
		return h[:]
	}
	return key
}

// createExchangeParties sets up issuer and holder accounts
func createExchangeParties() (*ExchangeParty, *ExchangeParty) {
	fmt.Println("Setting up exchange parties...")

	// Create issuer party
	issuer := &ExchangeParty{
		name:        "University Issuer",
		credentials: make(map[string]*credential.VerifiableCredential),
	}

	issuerCfg := &account.Config{
		StorageKey:  generateStorageKey("demo_issuer"),
		StoragePath: "./demo_issuer_storage",
		Environment: account.TargetSandbox,
		LogLevel:    account.LogWarn,
	}

	issuerAccount, err := account.New(issuerCfg)
	if err != nil {
		log.Fatal("Failed to create issuer account:", err)
	}
	issuer.account = issuerAccount

	// Create holder party
	holder := &ExchangeParty{
		name:        "Student Holder",
		credentials: make(map[string]*credential.VerifiableCredential),
	}

	holderCfg := &account.Config{
		StorageKey:  generateStorageKey("demo_holder"),
		StoragePath: "./demo_holder_storage",
		Environment: account.TargetSandbox,
		LogLevel:    account.LogWarn,
	}

	holderAccount, err := account.New(holderCfg)
	if err != nil {
		log.Fatal("Failed to create holder account:", err)
	}
	holder.account = holderAccount

	issuerInbox, _ := issuer.account.InboxOpen()
	holderInbox, _ := holder.account.InboxOpen()

	fmt.Printf("Issuer: %s\n", issuerInbox.String())
	fmt.Printf("Holder: %s\n", holderInbox.String())
	fmt.Println()

	return issuer, holder
}

// issueCredentialsForExchange creates credentials following issuance patterns
func issueCredentialsForExchange(issuer, holder *ExchangeParty) {
	fmt.Println("Issuing credentials...")

	issuerAddress, _ := issuer.account.InboxOpen()
	holderAddress, _ := holder.account.InboxOpen()

	// Issue email credential
	emailCred := createEmailCredential(issuer.account, issuerAddress, holderAddress)
	if emailCred != nil {
		holder.credentials["email"] = emailCred
		fmt.Println("✅ Email credential issued")
	}

	// Issue student ID credential
	studentCred := createStudentCredential(issuer.account, issuerAddress, holderAddress)
	if studentCred != nil {
		holder.credentials["student_id"] = studentCred
		fmt.Println("✅ Student ID credential issued")
	}

	// Issue degree credential
	degreeCred := createDegreeCredential(issuer.account, issuerAddress, holderAddress)
	if degreeCred != nil {
		holder.credentials["degree"] = degreeCred
		fmt.Println("✅ Degree credential issued")
	}

	fmt.Printf("Holder has %d credentials\n\n", len(holder.credentials))
}

// demonstrateExchangeWorkflow shows how issued credentials enable exchange
func demonstrateExchangeWorkflow(holder *ExchangeParty) {
	fmt.Println("Exchange workflow:")
	fmt.Println("Scenario: Employer requests proof of education")

	// Simulate exchange request
	requestedTypes := []string{"StudentCredential", "DegreeCredential"}
	fmt.Printf("Requested types: %v\n", requestedTypes)

	// Search credential store
	var matchingCreds []*credential.VerifiableCredential

	for credName, cred := range holder.credentials {
		credTypes := cred.CredentialType()
		for _, credType := range credTypes {
			for _, requestedType := range requestedTypes {
				if credType == requestedType {
					matchingCreds = append(matchingCreds, cred)
					fmt.Printf("✅ Found matching: %s (%s)\n", credName, credType)
					break
				}
			}
		}
	}

	if len(matchingCreds) == 0 {
		fmt.Println("❌ No matching credentials found")
		return
	}

	// Create presentation
	presentation := createCredentialPresentation(matchingCreds)
	if presentation != nil {
		fmt.Printf("✅ Presentation created with %d credentials\n", len(matchingCreds))
		fmt.Println("✅ Exchange completed successfully")
	}
}

// createCredentialPresentation creates a verifiable presentation from credentials
func createCredentialPresentation(credentials []*credential.VerifiableCredential) *credential.VerifiablePresentation {
	if len(credentials) == 0 {
		return nil
	}

	// For demo purposes, simulate successful presentation creation
	// In production, you would create a proper verifiable presentation
	// with the holder's signature
	return &credential.VerifiablePresentation{}
}

// Credential creation functions (following issuance example patterns)

func createEmailCredential(issuerAccount *account.Account, issuerAddress, holderAddress *signing.PublicKey) *credential.VerifiableCredential {
	claims := map[string]interface{}{
		"emailAddress": "student@university.edu",
		"verified":     true,
		"domain":       "university.edu",
	}

	credentialBuilder := credential.NewCredential().
		CredentialType(credential.CredentialTypeEmail).
		CredentialSubject(credential.AddressKey(holderAddress)).
		CredentialSubjectClaims(claims).
		Issuer(credential.AddressKey(issuerAddress)).
		ValidFrom(time.Now()).
		SignWith(issuerAddress, time.Now())

	unsignedCredential, err := credentialBuilder.Finish()
	if err != nil {
		log.Printf("Failed to build email credential: %v", err)
		return nil
	}

	emailCredential, err := issuerAccount.CredentialIssue(unsignedCredential)
	if err != nil {
		log.Printf("Failed to issue email credential: %v", err)
		return nil
	}

	return emailCredential
}

func createStudentCredential(issuerAccount *account.Account, issuerAddress, holderAddress *signing.PublicKey) *credential.VerifiableCredential {
	claims := map[string]interface{}{
		"studentId":      "STU-2024-001",
		"enrollmentDate": "2020-09-01",
		"status":         "enrolled",
		"program":        "Computer Science",
		"level":          "undergraduate",
	}

	credentialBuilder := credential.NewCredential().
		CredentialType([]string{"VerifiableCredential", "StudentCredential"}).
		CredentialSubject(credential.AddressKey(holderAddress)).
		CredentialSubjectClaims(claims).
		Issuer(credential.AddressKey(issuerAddress)).
		ValidFrom(time.Now()).
		SignWith(issuerAddress, time.Now())

	unsignedCredential, err := credentialBuilder.Finish()
	if err != nil {
		log.Printf("Failed to build student credential: %v", err)
		return nil
	}

	studentCredential, err := issuerAccount.CredentialIssue(unsignedCredential)
	if err != nil {
		log.Printf("Failed to issue student credential: %v", err)
		return nil
	}

	return studentCredential
}

func createDegreeCredential(issuerAccount *account.Account, issuerAddress, holderAddress *signing.PublicKey) *credential.VerifiableCredential {
	claims := map[string]interface{}{
		"degree":         "Bachelor of Science",
		"major":          "Computer Science",
		"graduationDate": "2024-05-15",
		"gpa":            3.8,
		"honors":         "magna cum laude",
		"institution":    "University of Technology",
	}

	credentialBuilder := credential.NewCredential().
		CredentialType([]string{"VerifiableCredential", "DegreeCredential"}).
		CredentialSubject(credential.AddressKey(holderAddress)).
		CredentialSubjectClaims(claims).
		Issuer(credential.AddressKey(issuerAddress)).
		ValidFrom(time.Now()).
		SignWith(issuerAddress, time.Now())

	unsignedCredential, err := credentialBuilder.Finish()
	if err != nil {
		log.Printf("Failed to build degree credential: %v", err)
		return nil
	}

	degreeCredential, err := issuerAccount.CredentialIssue(unsignedCredential)
	if err != nil {
		log.Printf("Failed to issue degree credential: %v", err)
		return nil
	}

	return degreeCredential
}

File: examples/server/02_credentials/02_exchanging_credentials/presentation_request/java/Main.kt
import com.joinself.selfsdk.kmp.account.Account
import com.joinself.selfsdk.kmp.credential.Address
import com.joinself.selfsdk.kmp.credential.CredentialBuilder
import com.joinself.selfsdk.kmp.credential.PresentationBuilder
import com.joinself.selfsdk.kmp.credential.VerifiableCredential
import com.joinself.selfsdk.kmp.credential.VerifiablePresentation
import com.joinself.selfsdk.kmp.keypair.signing.PublicKey
import com.joinself.selfsdk.kmp.time.Timestamp

/**
 * Represents a party in the credential exchange, holding an account and a
 * collection of their credentials.
 *
 * @param name A descriptive name for the party (e.g., "University Issuer").
 * @param account The underlying Self SDK account for this party.
 * @param credentials A map storing the verifiable credentials held by this party.
 */
data class ExchangeParty(
    val name: String,
    val account: Account,
    val credentials: MutableMap<String, VerifiableCredential> = mutableMapOf()
)

/**
 * Main entry point for the credential exchange demonstration.
 *
 * This example shows the foundational workflow for credential exchange:
 * 1. An **Issuer** (a university) issues multiple credentials to a **Holder** (a student).
 * 2. A **Verifier** (an employer) requests proof of education from the Holder.
 * 3. The **Holder** finds the required credentials in their wallet and creates a
 *    **Verifiable Presentation** to share with the Verifier.
 *
 * This demonstrates how issued credentials become the building blocks for secure data sharing.
 */
fun main() {
    println("Credential Exchange Demo")
    println("========================")

    // 1. Create the parties for the exchange scenario (an issuer and a holder)
    val (issuer, holder) = createExchangeParties()

    // 2. The issuer creates and sends several credentials to the holder
    issueCredentialsForExchange(issuer, holder)

    // 3. A verifier requests specific credentials, and the holder prepares a presentation
    demonstrateExchangeWorkflow(holder)

    println("\n✅ Exchange demo completed!")
    println("Press enter to exit")
    readln()
}

/**
 * Sets up the issuer and holder accounts and wraps them in the ExchangeParty data class.
 */
fun createExchangeParties(): Pair<ExchangeParty, ExchangeParty> {
    println("\nSetting up exchange parties...")

    // Create issuer party
    val issuerAccount = Common.setupAccount(storagePath = "./exchange_issuer_storage")
    val issuer = ExchangeParty("University Issuer", issuerAccount)

    // Create holder party
    val holderAccount = Common.setupAccount(storagePath = "./exchange_holder_storage")
    val holder = ExchangeParty("Student Holder", holderAccount)

    println()
    Common.displayAccountInfo(issuer.account, issuer.name)
    Common.displayAccountInfo(holder.account, holder.name)

    return Pair(issuer, holder)
}

/**
 * Issues a set of credentials from the issuer to the holder, populating the holder's "wallet".
 */
fun issueCredentialsForExchange(issuer: ExchangeParty, holder: ExchangeParty) {
    println("\nIssuing credentials for the holder...")

    val issuerAddress = Common.openInbox(issuer.account) ?: return
    val holderAddress = Common.openInbox(holder.account) ?: return

    // Issue and store an email credential
    createEmailCredential(issuer.account, issuerAddress, holderAddress)?.let {
        holder.credentials["email"] = it
        println("✅ Email credential issued")
    }

    // Issue and store a student ID credential
    createStudentCredential(issuer.account, issuerAddress, holderAddress)?.let {
        holder.credentials["student_id"] = it
        println("✅ Student ID credential issued")
    }

    // Issue and store a degree credential
    createDegreeCredential(issuer.account, issuerAddress, holderAddress)?.let {
        holder.credentials["degree"] = it
        println("✅ Degree credential issued")
    }

    println("🎓 Holder now has ${holder.credentials.size} credentials.")
}

/**
 * Simulates a verifier requesting credentials and the holder creating a presentation in response.
 */
fun demonstrateExchangeWorkflow(holder: ExchangeParty) {
    println("\nDemonstrating exchange workflow...")
    println("Scenario: An employer requests proof of education from the student holder.")

    // 1. A verifier requests credentials of specific types
    val requestedTypes = listOf("StudentCredential", "DegreeCredential")
    println("🔍 Verifier is requesting credential types: $requestedTypes")

    // 2. The holder searches their local credential store for matches
    val matchingCreds = holder.credentials.values.filter { cred ->
        val credTypes = cred.credentialType()
        // Check if any of the credential's types are in the requested list
        credTypes.any { it in requestedTypes }
    }

    if (matchingCreds.isEmpty()) {
        println("❌ No matching credentials found in holder's wallet.")
        return
    }

    println("✅ Found ${matchingCreds.size} matching credentials:")
    matchingCreds.forEach { cred ->
        // Find which of the credential's types was the one that matched
        val matchedType = cred.credentialType().first { it in requestedTypes }
        val identifier = cred.credentialSubjectClaims()["studentId"] ?: cred.credentialSubjectClaims()["degree"]
        println("   - '$identifier' ($matchedType)")
    }

    // 3. The holder creates a Verifiable Presentation containing the matching credentials
    val presentation = createCredentialPresentation(holder, matchingCreds)
    if (presentation != null) {
        println("✅ Presentation created with ${presentation.credentials().size} credentials.")
        println("   - Presentation Type: ${presentation.presentationType().toList()}")
        println("   - Holder: ${presentation.holder().address().encodeHex()}")
        println("✅ Exchange workflow completed successfully!")
    }
}

/**
 * Creates a Verifiable Presentation from a list of credentials, signed by the holder.
 * This is the standard way to package and share credentials with a verifier.
 */
fun createCredentialPresentation(holder: ExchangeParty, credentials: List<VerifiableCredential>): VerifiablePresentation? {
    if (credentials.isEmpty()) return null

    println("✍️  Creating Verifiable Presentation...")
    val holderAddress = Common.openInbox(holder.account) ?: return null

    return runCatching {
        val builder = PresentationBuilder()
            .presentationType(arrayOf("VerifiablePresentation", "ProofOfEducation"))
            .holder(Address.key(holderAddress))

        // Add all matching credentials to the presentation
        credentials.forEach { builder.credentialAdd(it) }

        val unsignedPresentation = builder.finish()

        // The holder's account signs and issues the presentation
        holder.account.presentationIssue(unsignedPresentation)
    }.getOrElse { error ->
        println("❌ Failed to create presentation: ${error.message}")
        null
    }
}

// --- Credential Creation Helper Functions ---

fun createEmailCredential(issuerAccount: Account, issuerAddress: PublicKey, holderAddress: PublicKey): VerifiableCredential? {
    return runCatching {
        val claims = mapOf(
            "emailAddress" to "student@university.edu",
            "verified" to true,
            "domain" to "university.edu"
        )

        val unsignedCredential = CredentialBuilder()
            .credentialType(arrayOf("VerifiableCredential", "EmailCredential"))
            .credentialSubject(Address.key(holderAddress))
            .credentialSubjectClaims(claims)
            .issuer(Address.key(issuerAddress))
            .validFrom(Timestamp.now())
            .signWith(issuerAddress, Timestamp.now())
            .finish()

        issuerAccount.credentialIssue(unsignedCredential)
    }.getOrElse {
        println("❌ Failed to issue email credential: ${it.message}")
        null
    }
}

fun createStudentCredential(issuerAccount: Account, issuerAddress: PublicKey, holderAddress: PublicKey): VerifiableCredential? {
    return runCatching {
        val claims = mapOf(
            "studentId" to "STU-2024-001",
            "enrollmentDate" to "2020-09-01",
            "status" to "enrolled",
            "program" to "Computer Science",
            "level" to "undergraduate"
        )

        val unsignedCredential = CredentialBuilder()
            .credentialType(arrayOf("VerifiableCredential", "StudentCredential"))
            .credentialSubject(Address.key(holderAddress))
            .credentialSubjectClaims(claims)
            .issuer(Address.key(issuerAddress))
            .validFrom(Timestamp.now())
            .signWith(issuerAddress, Timestamp.now())
            .finish()

        issuerAccount.credentialIssue(unsignedCredential)
    }.getOrElse {
        println("❌ Failed to issue student credential: ${it.message}")
        null
    }
}

fun createDegreeCredential(issuerAccount: Account, issuerAddress: PublicKey, holderAddress: PublicKey): VerifiableCredential? {
    return runCatching {
        val claims = mapOf(
            "degree" to "Bachelor of Science",
            "major" to "Computer Science",
            "graduationDate" to "2024-05-15",
            "gpa" to 3.8,
            "honors" to "magna cum laude",
            "institution" to "University of Technology"
        )

        val unsignedCredential = CredentialBuilder()
            .credentialType(arrayOf("VerifiableCredential", "DegreeCredential"))
            .credentialSubject(Address.key(holderAddress))
            .credentialSubjectClaims(claims)
            .issuer(Address.key(issuerAddress))
            .validFrom(Timestamp.now())
            .signWith(issuerAddress, Timestamp.now())
            .finish()

        issuerAccount.credentialIssue(unsignedCredential)
    }.getOrElse {
        println("❌ Failed to issue degree credential: ${it.message}")
        null
    }
}
